### PR TITLE
Updated product & customer models

### PR DESCRIPTION
**Pull request by prafulrathore**

### CODE REVIEW

File: subscription/views.py [hunk 1]
```diff
@@ -109,8 +134,8 @@ def get(self, request):
             # Invalid signature
             return HttpResponse(status=400)
         # Handle the checkout.session.completed event
-        if event['type'] == 'checkout.session.completed':
-            session = event['data']['object']
+        if event["type"] == "checkout.session.completed":
+            session = event["data"]["object"]
 
             return HttpResponse(status=200)
```
Comment (konnerdave) [inline subscription/views.py]: Use Rest Framework Response with Rest Framework status!

File: subscription/views.py [hunk 2]
```diff
@@ -109,8 +134,8 @@ def get(self, request):
...
         return HttpResponse(status=200)
```
Comment (konnerdave) [inline subscription/views.py]: Use Rest Framework Response with Rest Framework status!

File: subscription/views.py [hunk 3]
```diff
@@ -109,8 +134,8 @@ def get(self, request):
             # Invalid signature
             return HttpResponse(status=400)
         # Handle the checkout.session.completed event
-        if event['type'] == 'checkout.session.completed':
-            session = event['data']['object']
+        if event["type"] == "checkout.session.completed":
```
Comment (konnerdave) [inline subscription/views.py:136]: Just like session complete event, there is another event trigger after you cancel a subscription. Why are you not handling that?

File: subscription/views.py [hunk 4]
```diff
@@ -1,106 +1,131 @@
...
         try:
-            sub_id = request.user.customer.stripe_subscription_id if request.user else ""
+            # Get the customer subscription id
+            sub_id = request.user.customer.subscription_id if request.user else ""
+            # Delete the Subscription from the Stripe
             stripe.Subscription.delete(sub_id)
-            Customer.objects.filter(stripe_subscription_id=sub_id).delete()
-            Product.objects.filter(subscription_id=sub_id).delete()
+            # Delete the Customer instance
+            instance = Customer.objects.get(subscription_id=sub_id)
```
Comment (konnerdave) [inline subscription/views.py]: How can you be so sure that the subscription you wanted to delete has been deleted? Don't you think this should be done after you get feedback from the webhook?

Comment (konnerdave) [inline subscription/views.py]: Further you should delete subscription from the local subscription model and not the customer!

File: subscription/views.py [hunk 5]
```diff
@@ -1,106 +1,131 @@
...
+        # Get the Stripe subscription
+        subscription = stripe.Subscription.retrieve(session.subscription)
+        # Get the Stripe Product
+        plan = stripe.Product.retrieve(subscription.plan.product)
+        try:
+            # Get the product if its plan exist
+            product = Product.objects.get(product_id=plan.id)
+        except Product.DoesNotExist:
+            # Create a product if doesn't exist
+            product = Product.objects.create(
```
Comment (konnerdave) [inline subscription/views.py]: Why a new product is being created from a subscription? Subscription creation should rely on a product and not vice versa!

File: subscription/views.py [hunk 6]
```diff
@@ -1,106 +1,131 @@
...
 
         except Exception as e:
-            context = JsonResponse({'error': str(e)})
+            context = JsonResponse({"error": str(e)})
         return HttpResponse(context)
 
 
 class SuccessView(View):
+    """
+    Get the Success checkout for the subscription.
```
Comment (konnerdave) [inline subscription/views.py]: Can't understand the meaning of your docstring! Change this to something meaningful and straightforward!

File: subscription/views.py [hunk 7]
```diff
@@ -1,106 +1,131 @@
...
+                mode="subscription",
+                subscription_data={"trial_period_days": 7},
+                line_items=[{"price": settings.STRIPE_PRICE_ID, "quantity": 1}],
             )
-            context = JsonResponse({'sessionId': checkout_session['id']})
+            context = JsonResponse({"sessionId": checkout_session["id"]})
 
         except Exception as e:
-            context = JsonResponse({'error': str(e)})
+            context = JsonResponse({"error": str(e)})
```
Comment (konnerdave) [inline subscription/views.py]: Use Rest Framework response or you can also change this to an API's also as the response is JSON!

File: subscription/views.py [hunk 8]
```diff
@@ -1,106 +1,131 @@
...
-class CreateCheckoutSessionView(View):
-    @method_decorator(login_required)
+class CreateCheckoutSessionView(LoginRequiredMixin, View):
+    """
+    Create the checkout session for the subscription.
+    """
+
     def get(self, request):
-        domain_url = 'http://localhost:8000/'
+        domain_url = "http://localhost:8000/"
```
Comment (konnerdave) [inline subscription/views.py]: Get this from env or settings! The best idea is to get this dynamically from request! Like we get the set fro m`get_current_site`! Think about it!

File: subscription/views.py [hunk 9]
```diff
@@ -1,106 +1,131 @@
...
 from subscription.models import Customer, Product
 
 stripe.api_key = settings.STRIPE_SECRET_KEY
 
 
-class StripeConfiguration(View):
-    @method_decorator(login_required)
+class StripeConfiguration(LoginRequiredMixin, View):
+    """
+    Get the sesssion id based on Stripe primary key.
```
Comment (konnerdave) [inline subscription/views.py]: What is a stripe prmary key?

File: subscription/views.py [hunk 10]
```diff
@@ -1,106 +1,131 @@
 import stripe
 
 from django.conf import settings
-from django.contrib.auth.decorators import login_required
-from django.contrib.auth.models import User
-from django.contrib.auth.decorators import login_required
+from django.contrib.auth.mixins import LoginRequiredMixin
 from django.http import HttpResponse, JsonResponse
```
Comment (konnerdave) [inline subscription/views.py]: Swap them with Rest Framework response. Do some research about which views can be converted to API views!

File: subscription/models.py [hunk 11]
```diff
@@ -2,33 +2,34 @@
...
+
     def save(self, *args, **kwargs):
         self.price = self.price / 100
         super(Product, self).save(*args, **kwargs)
 
+
+class Customer(models.Model):
+    user = models.OneToOneField(to=User, on_delete=models.CASCADE)
+    customer_id = models.CharField(max_length=200, default="")
+    subscription_id = models.CharField(max_length=200, default="")
```
Comment (konnerdave) [inline subscription/models.py]: Shift this to a different model named Subscription.

File: subscription/models.py [hunk 12]
```diff
@@ -2,33 +2,34 @@
...
-    subscription_id = models.CharField(max_length=400, default="")
+    product_id = models.CharField(max_length=200, default="")
 
-    class meta:
+    class Meta:
         verbose_name = "Product"
         verbose_name_plural = "Products"
 
+    def __str__(self):
+        return f"Price of {self.name} is {self.price} {self.currency}"
```
Comment (konnerdave) [inline subscription/models.py]: Str method shouldn't be this big. Make it small! Furthermore only name should be used in str not the combination of every attribute.

File: subscription/models.py [hunk 13]
```diff
@@ -2,33 +2,34 @@
...
-        return f"{self.user.username}"
-
-
 class Product(models.Model):
     name = models.CharField(max_length=200)
     description = models.CharField(max_length=100, default="")
     price = models.FloatField()
     currency = models.CharField(max_length=10, default="")
-    subscription_id = models.CharField(max_length=400, default="")
+    product_id = models.CharField(max_length=200, default="")
```
Comment (konnerdave) [inline subscription/models.py]: Use the standard limit for CharField throughout the project which is 256.

File: subscription/models.py [hunk 14]
```diff
@@ -2,33 +2,34 @@
...
-    class meta:
-        verbose_name = "Customer"
-        verbose_name_plural = "Customers"
-
-    def __str__(self):
-        return f"{self.user.username}"
-
-
 class Product(models.Model):
     name = models.CharField(max_length=200)
```
Comment (konnerdave) [inline subscription/models.py]: Make the limit 256 and Shift the name attribute to an abstract model.

File: requirement.txt [hunk 15]
```diff
@@ -0,0 +1,37 @@
+appdirs==1.4.4
```
Comment (konnerdave) [inline requirement.txt:1]: A good idea is to put packages here at the time of installation. That way you will only be mentioning packages which are absolutely necessary and not the dependent ones. pip automatically fetches the dependent one. When you do pip freeze > requirements.txt, it puts all the packages here which becomes a lil confusing to manage.

File: core/views.py [hunk 16]
```diff
@@ -1,32 +1,61 @@
...
+    template_name = "core/django_registration/signup_complete.html"
+
+
+class HomeView(TemplateView):
+    template_name = "core/home.html"
+
+    def get_context_data(self, *args, **kwargs):
+        context = super(HomeView, self).get_context_data(*args, **kwargs)
+
+        if self.request.user.is_authenticated:
```
Comment (konnerdave) [inline core/views.py]: Does this check even matters? I don't see any separate authentication based step below! Remove this!

File: core/views.py [hunk 17]
```diff
@@ -1,32 +1,61 @@
...
 stripe.api_key = settings.STRIPE_SECRET_KEY
 
 
-class HomeView(View):
-    def get(self, request):
-        if not request.user.is_authenticated:
-            return render(request, 'core/home.html')
-        else:
+class Login(LoginView):
+    template_name = "core/registration/login.html"
```
Comment (konnerdave) [inline core/views.py]: Just for the sake of managing templates in one folder we are rewriting this again. It's not ideal, use the inbuilt ones!

File: core/views.py [hunk 18]
```diff
@@ -1,32 +1,61 @@
+from django.views.generic.base import TemplateView
```
Comment (konnerdave) [inline core/views.py]: Why is this import at the top?

File: core/templates/core/registration/login.html [hunk 19]
```diff
@@ -0,0 +1,17 @@
+{% extends 'core/base.html' %} {% load crispy_forms_tags %} {% block content %}
+<div class="card">
+  <div class="card-body">
+    <h2 class="text-center">Login</h2>
+    <form method="post">
+      {% csrf_token %} {{ form | crispy }}
+      <button
+        class="btn btn-outline-success"
+        style="margin-top: 6px; margin-left: 80px"
```
Comment (konnerdave) [inline core/templates/core/registration/login.html]: Never give inline styles, shift this to a CSS file!

File: core/templates/core/registration/login.html [hunk 20]
```diff
@@ -0,0 +1,17 @@
+{% extends 'core/base.html' %} {% load crispy_forms_tags %} {% block content %}
```
Comment (konnerdave) [inline core/templates/core/registration/login.html]: Write these in 3 separate lines!

File: core/templates/core/registration/login.html [hunk 21]
```diff
@@ -0,0 +1,17 @@
...
+<div class="card">
```
Comment (konnerdave) [inline core/templates/core/registration/login.html]: Leave a line or two after the above template tags.

File: core/templates/core/home.html [hunk 22]
```diff
@@ -1,5 +1,5 @@
-{% extends 'core/base.html' %} 
-{% load static %} 
+{% extends 'core/base.html' %}
+{% load static %}
 {% block content %}
 <!-- <header> -->
   <div class="card" style="width: 22rem;">
```
Comment (konnerdave) [inline core/templates/core/home.html]: Align left!

File: config/urls.py [hunk 23]
```diff
@@ -18,9 +18,9 @@
 
 urlpatterns = [
     # To the Admin Url
```
Comment (konnerdave) [inline config/urls.py]: Change this to `Admin URL`

File: config/urls.py [hunk 24]
```diff
@@ -18,9 +18,9 @@
...
-    path('admin/', admin.site.urls),
+    path("admin/", admin.site.urls),
     # To the Core Url
```
Comment (konnerdave) [inline config/urls.py]: Change this to `Core URLS`. Although it doesn' matter much as it's pretty much obvious.

File: config/settings.py [hunk 25]
```diff
@@ -34,62 +34,60 @@
 
 INSTALLED_APPS = [
```
Comment (konnerdave) [inline config/settings.py]: You can divide this list into three:-
1. DJANGO_APPS = [ ]
2. PACKAGE_APPS = [ ]
3. CUSTOM_APPS = [ ]
then,
INSTALLED_APPS = DJANGO_APPS +  PACKAGE_APPS + CUSTOM_APPS

**Review by konnerdave — CHANGES_REQUESTED** [review]

**Comment by prafulrathore** [comment]
Thank you for telling me , i will do!

**Pull request merged**